File: app/views/test/web_component/targets.rb
# frozen_string_literal: true

module Views
  class Test::WebComponent::Targets < Application
    include Phlexible::Rails::AutoLayout

    sideload_assets js: { type: 'module' }

    register_element :pui_test_targets

    self.page_title = 'WebComponent — targets'

    def view_template
      h1 { 'Targets and selector helpers' }

      pui_test_targets(id: 'targets-host') do
        div(data: { target: 'menu' }, id: 't1') { 'first' }
        div(data: { target: 'menu' }, id: 't2') { 'second' }
        div(data: { target: 'sole' }, id: 'sole') { 'sole' }
        button(id: 'count-btn', 'on-click': 'reportCounts') { 'count' }
        button(id: 'override-btn', 'on-click': 'reportOverride') { 'override' }
        button(id: 'q-btn', 'on-click': 'reportQueries') { 'q' }
        ul(id: 'targets-log')
      end
    end
  end
end

File: app/views/test/web_component/targets.js
import WebComponent from '../../../../lib/proscenium/ui/web_component'

class TestTargets extends WebComponent {
  static componentName = 'pui-test-targets'
  static actions = {
    click: ['reportCounts', 'reportOverride', 'reportQueries']
  }

  #log(line) {
    const li = document.createElement('li')
    li.textContent = line
    this.querySelector('#targets-log').appendChild(li)
  }

  reportCounts() {
    const targets = this.$menuTargets
    this.#log(
      `menu:${targets.length}:${Array.isArray(targets)}:${this.$menuTarget?.id}:${this.$soleTarget?.id}`
    )
  }

  reportOverride() {
    // Override the target lookup with an explicit instance property — Proxy must yield to it.
    this.$menuTarget = { id: 'overridden' }
    this.#log(`override:${this.$menuTarget.id}`)
    delete this.$menuTarget
    this.#log(`restore:${this.$menuTarget.id}`)
  }

  reportQueries() {
    // $ and $$ helpers
    const single = this.$('#sole')
    const all = this.$$('[data-target="menu"]')
    this.#log(`q:${single?.id}:${all.length}:${Array.isArray(all)}`)
  }
}

WebComponent.register(TestTargets)
